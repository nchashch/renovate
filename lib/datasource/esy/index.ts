import { logger } from '../../logger';
import got from '../../util/got';
import { PkgReleaseConfig, ReleaseResult, Release } from '../common';

export async function getPkgReleases({
  lookupName,
}: PkgReleaseConfig): Promise<ReleaseResult> {
  const path: string = `/repos/ocaml/opam-repository/contents/packages/${lookupName}/`;
  const baseUrl: string = 'https://api.github.com';
  const packageUrl: string = baseUrl + path;
  console.log(packageUrl);
  let res: any = await got(packageUrl, {
    hostType: 'esy',
    json: true,
  });
  // TODO: Read homepage, name, and maybe other metadata from /packages/${lookupName}/${lookupName.1.2.3}/opam
  // where 1.2.3 can be any existing version, check if there is a mechanism for deprication
  const versions: string[] = res.body.map(file =>
    file.name.substring(file.name.indexOf('.') + 1)
  );
  const releases: Release[] = versions.map(version => ({ version }));
  const result: ReleaseResult = { releases };
  return result;
}
