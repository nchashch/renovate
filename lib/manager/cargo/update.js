const { getDep, getInlineTableDep, getTableDep } = require('./extract');

module.exports = {
  updateDependency,
};

// FIXME: Currently this function just replaces currentValue with newValue at
//        number line without checking if it is the right dependency
// TODO: Add more tests
function updateDependency(currentFileContent, upgrade) {
  logger.debug({ config: upgrade }, 'cargo.updateDependency()');
  const { depName, depType, lineNumber, newValue } = upgrade;
  const lines = currentFileContent.split('\n');
  if (depType === 'normal') {
    const line = lines[lineNumber];
    const currentValue = getDep(line).currentValue;
    lines[lineNumber] = line.replace(currentValue, newValue);
  } else if (depType == 'inlineTable') {
    const line = lines[lineNumber];
    const currentValue = getInlineTableDep(line).currentValue;
    lines[lineNumber] = line.replace(currentValue, newValue);
  } else if (depType == 'standardTable') {
    const match = line.match('dependencies[.]');
    const dep = getTableDep(match, lines, lineNumber);
    const currentValue = dep.currentValue;
    const versionLineNumber = dep.versionLineNumber;
    lines[versionLineNumber] = lines[versionLineNumber].replace(currentValue, newValue);
  }
  return lines.join('\n');
}
