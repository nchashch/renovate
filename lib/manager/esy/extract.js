module.exports = {
  extractPackageFile,
};

// NOTE: It looks like esy uses the same dependencies, devDependencies,
// buildDependencies fields in package.json as npm. But package.json also can
// have esy field. Also it uses both npm and opam, so a new datasource might
// be necessary.
function extractPackageFile(content) {
  logger.trace(`extractPackageFile()`);
  try {
    const doc = JSON.parse(content);
    if (doc.esy) {
      logger.info('esy field is present');
    }
  } catch (err) {
    return {};
  }
  return {};
}
