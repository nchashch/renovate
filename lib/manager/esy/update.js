module.exports = {
  updateDependency,
};

function updateDependency(fileContent, upgrade) {
  logger.trace(`updateDependency(${upgrade.depName})`);
  return fileContent;
}
