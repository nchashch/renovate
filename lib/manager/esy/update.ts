import { isEqual } from 'lodash';
import { logger } from '../../logger';
import { Upgrade } from '../common';

export function updateDependency(fileContent: string, upgrade: Upgrade) {
  logger.trace(`updateDependency(${upgrade.depName})`);
  const { depType, depName, newValue, currentValue } = upgrade;
  try {
    const parsedContents = JSON.parse(fileContent);
    const oldVersion = parsedContents[depType][depName];
    if (oldVersion !== currentValue) {
      logger.warn(
        `Actual value in ${depType}.${depName} = ${oldVersion} is not equal to currentValue = ${currentValue}`
      );
    }
    if (oldVersion === newValue) {
      logger.trace('Version is already updated');
      return fileContent;
    }
    parsedContents[depType][depName] = newValue;
    const searchString = `"${oldVersion}"`;
    const newString = `"${newValue}"`;
    let newFileContent = null;
    let searchIndex = fileContent.indexOf(`"${depType}"` + depType.length);
    logger.trace(`Starting search at index ${searchIndex}`);
    // Iterate through the rest of the file
    for (; searchIndex < fileContent.length; searchIndex += 1) {
      // First check if we have a hit for the old version
      if (matchAt(fileContent, searchIndex, searchString)) {
        logger.trace(`Found match at index ${searchIndex}`);
        // Now test if the result matches
        const testContent = replaceAt(
          fileContent,
          searchIndex,
          searchString,
          newString
        );
        // Compare the parsed JSON structure of old and new
        if (isEqual(parsedContents, JSON.parse(testContent))) {
          newFileContent = testContent;
          break;
        }
      }
    }
    if (!newFileContent) {
      logger.info(
        { fileContent, parsedContents, depType, depName, newValue },
        'Warning: updateDependency error'
      );
      return fileContent;
    }
    return newFileContent;
  } catch (err) {
    logger.info({ err }, 'updateDependency error');
    return null;
  }
}

// Return true if the match string is found at index in content
function matchAt(content: string, index: number, match: string): boolean {
  return content.substring(index, index + match.length) === match;
}

// Replace oldString with newString at location index of content
function replaceAt(
  content: string,
  index: number,
  oldString: string,
  newString: string
): string {
  logger.debug(`Replacing ${oldString} with ${newString} at index ${index}`);
  return (
    content.substr(0, index) +
    newString +
    content.substr(index + oldString.length)
  );
}
