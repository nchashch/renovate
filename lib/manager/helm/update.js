module.exports = {
  updateDependency,
};

function updateDependency(fileContent, upgrade) {
  return fileContent;
}
