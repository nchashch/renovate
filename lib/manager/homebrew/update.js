const got = require('got');
const crypto = require('crypto');
const semver = require('../../versioning/semver');
const { parseUrlPath } = require('./extract');
const { skip, isSpace, removeComments } = require('./util');

module.exports = {
  updateDependency,
};

async function updateDependency(content, upgrade) {
  /*
    1. Update url field
    2. Update sha256 field
   */
  let newContent = content;
  let newUrl;
  let file;
  // Example urls:
  // "https://github.com/bazelbuild/bazel-watcher/archive/v0.8.2.tar.gz"
  // "https://github.com/aide/aide/releases/download/v0.16.1/aide-0.16.1.tar.gz"
  try {
    newUrl = `https://github.com/${upgrade.ownerName}/${
      upgrade.repoName
    }/releases/download/${upgrade.newValue}/${upgrade.repoName}-${semver.coerce(
      upgrade.newValue
    )}.tar.gz`;
    file = (await got(newUrl, { encoding: null })).body;
  } catch (err) {
    logger.debug(
      'Failed to download release download - trying archive instead'
    );
    newUrl = `https://github.com/${upgrade.ownerName}/${
      upgrade.repoName
    }/archive/${upgrade.newValue}.tar.gz`;
    file = (await got(newUrl, { encoding: null })).body;
  }
  const newSha256 = crypto
    .createHash('sha256')
    .update(file)
    .digest('hex');
  if (upgrade.newValue !== parseUrlPath(newUrl).currentValue) {
    logger.debug(`Failed to update url for dependency ${upgrade.depName}`);
    return content;
  }
  newContent = updateUrl(content, newUrl);
  newContent = newContent.replace(upgrade.sha256, newSha256);
  return newContent;
}

function updateUrl(content, url) {
  const urlRegExp = /(^|\s)url(\s)/;
  let i = content.search(urlRegExp);
  if (isSpace(content[i])) {
    i += 1;
  }
  let newContent = replaceUrl(i, content, url);
  const testContent = getTestContent(content, url);
  while (removeComments(newContent) !== testContent) {
    i += 'url'.length;
    i += content.substring(i).search(urlRegExp);
    if (isSpace(content[i])) {
      i += 1;
    }
    newContent = replaceUrl(i, content, url);
  }
  return newContent;
}

function getTestContent(content, url) {
  const urlRegExp = /(^|\s)url(\s)/;
  const cleanContent = removeComments(content);
  let j = cleanContent.search(urlRegExp);
  if (isSpace(cleanContent[j])) {
    j += 1;
  }
  const testContent = replaceUrl(j, cleanContent, url);
  return testContent;
}

// FIXME: Copy and pasted from parseUrl function in extract.js (refactor)
function replaceUrl(idx, content, url) {
  let i = idx;
  i += 'url'.length;
  i = skip(i, content, c => {
    return isSpace(c);
  });
  const chr = content[i];
  if (chr !== '"' && chr !== "'") {
    return null;
  }
  i += 1;
  let j = i;
  j = skip(i, content, c => {
    return c !== '"' && c !== "'" && !isSpace(c);
  });
  const newContent = content.substring(0, i) + url + content.substring(j);
  return newContent;
}
