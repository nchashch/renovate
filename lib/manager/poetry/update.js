module.exports = {
  updateDependency,
};

function updateDependency(currentFileContent, upgrade) {
  logger.trace({ config: upgrade }, 'poetry.updateDependency()');
  const lines = currentFileContent.split('\n');
  return lines.join('\n');
}
