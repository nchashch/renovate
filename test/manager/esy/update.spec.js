import { readFileSync } from 'fs';
import { updateDependency } from '../../../lib/manager/esy/update';

const esy1json = readFileSync('test/manager/esy/_fixtures/esy.1.json', 'utf8');
const esy2json = readFileSync('test/manager/esy/_fixtures/esy.2.json', 'utf8');

describe('lib/manager/esy/update', () => {
  describe('updateDependency()', () => {
    it('returns null for empty upgrade', async () => {
      const content = 'some content';
      const upgrade = {};
      expect(await updateDependency(content, upgrade)).toBeNull();
    });
    it('updates a dependency correctly', async () => {
      const content = esy1json;
      const upgrade = {
        depType: 'dependencies',
        depName: 'ocaml',
        currentValue: '~4.6.0',
        newValue: '5.0.0',
      };
      const res = await updateDependency(content, upgrade);
      expect(res).not.toBeNull();
      expect(res).not.toMatch(content);
      expect(res).toMatchSnapshot();
    });
    it('updates a dependency correctly', async () => {
      const content = esy2json;
      const upgrade = {
        depType: 'dependencies',
        depName: 'ocaml',
        currentValue: '~4.6.0',
        newValue: '5.0.0',
      };
      const res = await updateDependency(content, upgrade);
      expect(res).not.toBeNull();
      expect(res).not.toMatch(content);
      expect(res).toMatchSnapshot();
    });
    it('returns unchanged content if newValue === currentValue ', async () => {
      const content = esy2json;
      const upgrade = {
        depType: 'dependencies',
        depName: 'ocaml',
        currentValue: '~4.6.0',
        newValue: '~4.6.0',
      };
      const res = await updateDependency(content, upgrade);
      expect(res).toMatch(content);
    });
    it('returns unchanged content if depName is not found', async () => {
      const content = `{
  "name": "hello-reason",
  "version": "0.1.0",
  "description": "Example Reason Esy Project",
  "license": "MIT",
  "dependencies": {
    "foo": "1.2.3"
  }
}`;
      const upgrade = {
        depType: 'dependencies',
        depName: 'will_not_be_found',
        currentValue: '1.2.4',
        newValue: '2.0.0',
      };
      const res = await updateDependency(content, upgrade);
      expect(res).not.toBeNull();
      expect(res).toMatch(content);
    });
    it('updates correctly if there are two dependencies with the same depName in different sections', async () => {
      const content = `{
  "name": "hello-reason",
  "version": "0.1.0",
  "description": "Example Reason Esy Project",
  "license": "MIT",
  "dependencies": {
    "foo": "1.2.3"
  },
  "devDependencies": {
    "foo": "1.2.4"
  }
}`;
      const upgrade = {
        depType: 'devDependencies',
        depName: 'foo',
        currentValue: '1.2.4',
        newValue: '2.0.0',
      };
      const res = await updateDependency(content, upgrade);
      expect(res).not.toBeNull();
      expect(res).not.toMatch(content);
      expect(res).toMatchSnapshot();
    });
    it('updates correctly if there are two dependencies with the same version in the same section', async () => {
      const content = `{
  "name": "hello-reason",
  "version": "0.1.0",
  "description": "Example Reason Esy Project",
  "license": "MIT",
  "dependencies": {
    "foo": "1.2.3",
    "bar": "1.2.3"
  }
}`;
      const upgrade = {
        depType: 'dependencies',
        depName: 'bar',
        currentValue: '1.2.3',
        newValue: '2.0.0',
      };
      const res = await updateDependency(content, upgrade);
      expect(res).not.toBeNull();
      expect(res).not.toMatch(content);
      expect(res).toMatchSnapshot();
    });
  });
});
