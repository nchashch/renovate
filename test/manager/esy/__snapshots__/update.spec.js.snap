// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/manager/esy/update updateDependency() updates a dependency correctly 1`] = `
"{
  \\"name\\": \\"hello-ocaml\\",
  \\"version\\": \\"0.1.0\\",
  \\"description\\": \\"OCaml workflow with Esy\\",
  \\"license\\": \\"MIT\\",
  \\"esy\\": {
    \\"build\\": \\"dune build -p #{self.name}\\",
    \\"release\\": {
      \\"bin\\": \\"hello\\",
      \\"includePackages\\": [
        \\"root\\",
        \\"@opam/camomile\\"
      ]
    }
  },
  \\"scripts\\": {
    \\"test\\": \\"esy x hello\\"
  },
  \\"dependencies\\": {
    \\"@opam/dune\\": \\"*\\",
    \\"@opam/lambda-term\\": \\"*\\",
    \\"@opam/lwt\\": \\"^4.0.0\\",
    \\"ocaml\\": \\"5.0.0\\"
  },
  \\"devDependencies\\": {
    \\"@opam/merlin\\": \\"^3.0.3\\",
    \\"ocaml\\": \\"~4.6.0\\"
  }
}
"
`;

exports[`lib/manager/esy/update updateDependency() updates a dependency correctly 2`] = `
"{
  \\"name\\": \\"hello-reason\\",
  \\"version\\": \\"0.1.0\\",
  \\"description\\": \\"Example Reason Esy Project\\",
  \\"license\\": \\"MIT\\",
  \\"esy\\": {
    \\"build\\": \\"refmterr dune build -p #{self.name}\\",
    \\"NOTE\\": \\"Optional release Section. Customizes result of \`esy release\`\\",
    \\"release\\": {
      \\"bin\\": [
        \\"Hello\\"
      ]
    },
    \\"buildEnv\\": {
      \\"ODOC_SYNTAX\\": \\"re\\"
    }
  },
  \\"scripts\\": {
    \\"test\\": \\"esy x Hello\\",
    \\"doc\\": \\"esy dune build @doc\\"
  },
  \\"dependencies\\": {
    \\"@opam/dune\\": \\"*\\",
    \\"@reason-native/console\\": \\"*\\",
    \\"@reason-native/pastel\\": \\"*\\",
    \\"@reason-native/rely\\": \\"*\\",
    \\"@esy-ocaml/reason\\": \\">= 3.4.0 < 3.6.0\\",
    \\"refmterr\\": \\"*\\",
    \\"ocaml\\": \\"5.0.0\\"
  },
  \\"devDependencies\\": {
    \\"@opam/merlin\\": \\"*\\",
    \\"ocaml\\": \\"~4.6.0\\",
    \\"@opam/odoc\\": \\"*\\"
  }
}
"
`;

exports[`lib/manager/esy/update updateDependency() updates correctly if there are two dependencies with the same depName in different sections 1`] = `
"{
  \\"name\\": \\"hello-reason\\",
  \\"version\\": \\"0.1.0\\",
  \\"description\\": \\"Example Reason Esy Project\\",
  \\"license\\": \\"MIT\\",
  \\"dependencies\\": {
    \\"foo\\": \\"1.2.3\\"
  },
  \\"devDependencies\\": {
    \\"foo\\": \\"2.0.0\\"
  }
}"
`;

exports[`lib/manager/esy/update updateDependency() updates correctly if there are two dependencies with the same version in the same section 1`] = `
"{
  \\"name\\": \\"hello-reason\\",
  \\"version\\": \\"0.1.0\\",
  \\"description\\": \\"Example Reason Esy Project\\",
  \\"license\\": \\"MIT\\",
  \\"dependencies\\": {
    \\"foo\\": \\"1.2.3\\",
    \\"bar\\": \\"2.0.0\\"
  }
}"
`;
